keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+with improved support for exclude matching.
+
+*Excluding*:
+
+Exclude support make it possible to do things like validate all namespaces
+except `kube-system`. This is difficult to support with out direct exclude
+support, particularly for 3rd party policy enforcements systems that cannot
+assume permission to set labels on kube-system.
+
+Exclude matching will be offered adding `exclude<fieldname>` for match fields
+where exclude is appropriate. Fields such as `namespaceSelector` that already
+offer exclusion (e.g. via the `NotIn` operator) do not need a corresponding
+`exclude<fieldname>`.
+
+E.g.:
+
+```yaml
+match:
+  excludeNamespaces: ["kube-system"] # excludeNamespaces and namespaces are mutually exclusive
+  excludePermissions: ["all-the-superpowers"]
+  namespaceSelector: # Already has exclude support via NotIn and DoesNotExist
+  - keys: "xyz"
+    operator: NotIn
+    values: ["1"]
+  excludeResourceRules: # excludeResourceRules takes precedent over resourceRules
+  - apiGroups:        ["apps"]
+    apiVersions:      ["*"]
+    operations:       ["*"]
+    resources:        ["deployments"]
+  resourceRules:
+  - apiGroups:        ["apps"]
+    apiVersions:      ["*"]
+    operations:       ["*"]
+    resources:        ["*"]
+  ...
+```
+
+*Special case: apiGroup + resource + operation matching*
+
+For admission webhooks, at least one `spec.rules` must be declared to state
+which apiGroup + resource + operations the webhook operates on. To configure a
+webhook to match everything (which is a very bad idea), a match rule would need
+to be written to state that, e.g.:
+
+```yaml
+spec:
+  rules:
+  - apiGroups:   ["*"]
+    apiVersions: ["*"]
+    operations:  ["*"]
+    resources:   ["*"]
+```
+
+This forces the webhook configuration author to explicitly declare what
+they intend to match.
+
+The same principle applies here but with one major difference-- if a policy
+definition has a match rules for apiGroup + resource + operation, then all
+bindings of that policy are already constrainted to that apiGroup + resource +
+operation match.
+
+Take, for example:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+...
+spec:
+  paramSource: ...
+  match:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+```
+
+Since this policy is constrainted to create/update of deployments. Policy
+bindings don't need to repeat this constraint. This would be sufficient:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: PolicyBinding
+...
+spec:
+  match:
+    namespaceSelectors:
+    - key: environment
+      operator: Exists
+ ...
+```
+
+We can enforce this by requiring:
+
+- Policy definitions match rules are validated to match one or more apiGroup +
+  resource + operation.
+- Policy bindings are not required to match apiGroup + resource +
```

Comment:
But they may? In cases where a policy definition matches more than one group/resource/operation and the binding author wants to further scope that down?

Reply:
+1 are they allowed to do this? I can't think of a good reason so an example would help.

Reply:
(now that I see @TristonianJones' comment) The matches are doing different things; the policy needs a match that defines resources, so that the CEL expressions can be evaluated for type safety. The binding doesn't need to (and IMO shouldn't be able to) repeat that. But it needs to decide based on metadata (name, namespace, labels, etc) whether a policy needs to be run on a given object.

Reply:
> are they allowed to do this?

For group/resource/operation, the way the KEP is written the binding could further restrict it down. There are practical applications for being able to do this.  E.g. for a rollout of a policy, being able pick exactly what resources will transition from warning to deny for a policy can be achieved if the policy bindings (which pick enforcement) can also pick which resources that enforcement should apply to.

+1 to @lavalamp's comment about type safety.